Implement a "purgatory" state for ranges failing to replicate.

spencerkimball opened the pull request:
A purgatory is where replicas which fail to up-replicate are placed
in order to be quickly processed again when new store information
is gossiped. The purgatory also offers a much less noisy logging
mechanism. Initial failures for any replica are logged as before,
but further failures aren't logged. As long as a queue has replicas
in purgatory, it periodically logs how many, grouped by the error
message. The default logging frequency of this purgatory reporting
is once every 10m.

The purgatory mechanism is part of the base queue, so we can use it
for other queue implementations as the need arises.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4172)

<!-- Reviewable:end -->

spencerkimball committed:
Implement a "purgatory" state for ranges failing to replicate.

A purgatory is where replicas which fail to up-replicate are placed
in order to be quickly processed again when new store information
is gossiped. The purgatory also offers a much less noisy logging
mechanism. Initial failures for any replica are logged as before,
but further failures aren't logged. As long as a queue has replicas
in purgatory, it periodically logs how many, grouped by the error
message. The default logging frequency of this purgatory reporting
is once every 10m.

The purgatory mechanism is part of the base queue, so we can use it
for other queue implementations as the need arises.

tamird commented:
LGTM

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9nzgTdmmE4-JMCPUl5:85191517) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
"a channel that signals when it's time"

---

*[storage/queue.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nyuL1gaQ20NHx4NkX:-K9nyuL1gaQ20NHx4NkY:-1262293497) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L162)):*
rearrange this so the comment can just be "protects fields below"

---

*[storage/queue.go, line 273 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzCAzWVa_rEtg58BK:-K9nzCAzWVa_rEtg58BL:134114623) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L273)):*
there's no need to nil check before map lookups http://play.golang.org/p/Cgvjs5SOud

---

*[storage/queue.go, line 365 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o--crgomEio7m7nJ3:-K9o--crgomEio7m7nJ4:-1234964601) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L365)):*
while you're here, could this set to nil?

---

*[storage/queue.go, line 428 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o-_EW_LTHgMVgSuie:-K9o-_EW_LTHgMVgSuif:812869107) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L428)):*
instead of making the stopper nil, can you use `RunTask` here?

---

*[storage/queue.go, line 456 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o07utC_iyIrybp9fE:-K9o-wjmZl364765oY1T:1898397783) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L456)):*
s/slice/map/

also, can't this overwrite an existing entry?

---

*[storage/queue.go, line 462 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o0AUJ9RvqvXKkBYUQ:-K9o0AUJ9RvqvXKkBYUR:1734848452) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L462)):*
purgatory

---

*[storage/queue.go, line 483 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o0Qk_OLBJNwQwKD_k:-K9o0Qk_OLBJNwQwKD_l:-1799865299) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L483)):*
should this long the elapsed time like the other site does? is it possible not to repeat the logging logic?

---

*[storage/queue_test.go, line 452 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o1oeI7FOD2i-xc0Ax:-K9o1oeJrUdmwrWA-z5a:1258997081) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue_test.go#L452)):*
broken comment

---

*[storage/queue_test.go, line 493 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o2EWK-cRROLkgdgcb:-K9o2EWK-cRROLkgdgcc:184565531) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue_test.go#L493)):*
this should also check that the priority queue is empty

---

*[storage/queue_test.go, line 517 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o2IEfmEo4C-vJ9_4L:-K9o2IEfmEo4C-vJ9_4M:184565531) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue_test.go#L517)):*
this should also check that the priority queue is empty

---

*[storage/queue_test.go, line 533 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o27xPY-3LJ-6kVRYB:-K9o27xPY-3LJ-6kVRYC:-903212105) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue_test.go#L533)):*
are no longer in purgatory

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-:-K9o2QNot_99ioCv2gxM:676834233)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 8 of 9 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 365 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o--crgomEio7m7nJ3:-K9o8kJ1G01pNK3mK4It:-1544852864) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L365)):*
Not sure that's safe depending on what else is going on in the system. There are numerous places where `MaybeAdd` is called directly on queues. Better to leave this as is I think.

---

*[storage/queue.go, line 428 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o-_EW_LTHgMVgSuie:-K9o9AzgU7kwURfERuP2:1414605971) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L428)):*
The problem is that from `DrainQueue`, which calls this method, I don't have a stopper.

---

*[storage/queue.go, line 456 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o07utC_iyIrybp9fE:-K9o9i8SjluQT_AW_c2_:-1547720688) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L456)):*
If the replica is already in purgatory, then we wouldn't have gotten this far, as it would have been in `bq.replicas` map and we'd have returned already.

---

*[storage/queue.go, line 483 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o0Qk_OLBJNwQwKD_k:-K9oAIEqyGwl60has5RQ:-868623274) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L483)):*
sure, i'll add that back in.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 365 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o--crgomEio7m7nJ3:-K9oFVLM1E3tBBOmGmHU:-1887860605) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L365)):*
Seems like those operations should be under a RunTask. Can you add a TODO?

---

*[storage/queue.go, line 428 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o-_EW_LTHgMVgSuie:-K9oF-wE4WkvpahSUB2A:-804639144) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L428)):*
Yeah, and there's also a lot of stopper passing around after this change. Is it possible put the stopper on the queue struct?

---

*[storage/queue.go, line 483 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o0Qk_OLBJNwQwKD_k:-K9oF9jWzGMllI_Fd3sm:-288668811) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L483)):*
Can you factor out the logging code?

---

*[storage/replicate_queue.go, line 60 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9oFONQp5GVE3mOMUlo:-K9oFONQp5GVE3mOMUlp:883066165) ([raw file](https://github.com/cockroachdb/cockroach/blob/6360f04d60321df94878887b072cce7dce9dc155/storage/replicate_queue.go#L60)):*
can you add a comment explaining when `g` might be nil?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 365 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o--crgomEio7m7nJ3:-K9oLTI0O78LoNS0lo7U:-966914839) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L365)):*
Do you ever get the feeling that too much of our complexity is centered around making tests shut down correctly? Sigh...

---

*[storage/queue.go, line 428 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o-_EW_LTHgMVgSuie:-K9oLbpefLXS2NmPldmQ:2029672679) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L428)):*
I suppose, though I'm going to try to rearrange things so there's less stopper passing instead.

---

*[storage/queue.go, line 483 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o0Qk_OLBJNwQwKD_k:-K9oM0mu9P7oZggyBCBK:-1814325345) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L483)):*
OK, will try to clean things up a bit with stopper etc.

---

*[storage/replicate_queue.go, line 60 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9oFONQp5GVE3mOMUlo:-K9oM4NBr1zeB4hu8euv:-1417167106) ([raw file](https://github.com/cockroachdb/cockroach/blob/6360f04d60321df94878887b072cce7dce9dc155/storage/replicate_queue.go#L60)):*
unittests obviously... will add comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Was this change motivated by something in particular (i.e. up-replication slugging along)?

---

Reviewed 6 of 9 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/gc_queue.go, line 442 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9pJfl5Lbfs0MVwIrHB:-K9pJfl5Lbfs0MVwIrHC:-1641979085) ([raw file](https://github.com/cockroachdb/cockroach/blob/6360f04d60321df94878887b072cce7dce9dc155/storage/gc_queue.go#L442)):*
that wording could also be interpreted as saying that not sending to purgatory means quicker retries, slightly misleading.

---

*[storage/queue.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nyuL1gaQ20NHx4NkX:-K9pIbhtFF8fEw__JIv-:235027556) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L162)):*
also, a good opportunity to intro the  anonymous-struct-with-lock-pattern here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-:-K9pKfAeIFTEEJMXiGzJ:1333270159)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Now that I'm running the server locally a lot I'm seeing all kinds of extremely annoying things which will be an order of magnitude more annoying at scale. The small irritation which prompted this change was excessive logging when jamming tons of data into a single node local server about how all of ranges were under replicated. After staring at hundreds of messages going by every minute announcing the same problem I decided that I could kill two birds with one stone.

As you say, the larger problem was that if you have a bunch of under replicated ranges, many of which you are well-aware of, there's no urgency to fix them once more capacity comes online.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nyuL1gaQ20NHx4NkX:-K9rxnD3sNQCL9r-udV0:-10818774) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L162)):*
I missed this new pattern. Could you point me at an example?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-:-K9rsXa19Tt5pf_OLcBJ:-817828873)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok. Some of that info could be put into a comment somewhere in the explanation of `purgatory` to illustrate the use cases.

---

Reviewed 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[storage/allocator.go, line 256 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s22pW9PpBnkFTxGlr:-K9s22pW9PpBnkFTxGls:-1429004594) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3df9481336ecaa8e56aa6c88172f96008ef3fca/storage/allocator.go#L256)):*
this error was previously better, and for the one a few lines up we should ideally include the number of nodes. That gets harder with this change - maybe just put a string on `allocatorErr` and create the message here to avoid the loss of clarity.

---

*[storage/queue.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nyuL1gaQ20NHx4NkX:-K9s1dTWykr68d7rWRrD:-46757541) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L162)):*
`(storage.Replica).mu` is a good one.

---

*[storage/queue.go, line 418 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s2QySRRS0RD_oseZr:-K9s2QySRRS0RD_oseZs:-1098716931) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L418)):*
no logging on error?

---

*[storage/queue.go, line 424 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s2XHMWITNUsk6auPO:-K9s2XHMWITNUsk6auPP:-1382405244) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L424)):*
That makes it sound like it's illegal to call with a non-retryable error, but that's not what you mean.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-:-K9s1mTbvFgf5MLXZTJ1:-1915952023)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[storage/allocator.go, line 128 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s2YDgS1EbRZ_4xFlS:-K9s2YDgS1EbRZ_4xFlT:-32614260) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/allocator.go#L128)):*
no need to bind the receiver

---

*[storage/allocator.go, line 133 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s2_1Iu7CYDudcUPbN:-K9s2_1Iu7CYDudcUPbO:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/allocator.go#L133)):*
ditto

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9s1C3xx8AorzXq6xSX:2004358416) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
still needs s/to/that/

---

*[storage/queue.go, line 482 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s4DI9Ocl3no_N0OWb:-K9s4DI9Ocl3no_N0OWc:-1396575290) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L482)):*
use `bq.eventLog`?

---

*[storage/queue.go, line 497 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s4KOWqVlgJmiNmLCL:-K9s4KOWqVlgJmiNmLCM:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L497)):*
ditto

---

*[storage/queue.go, line 538 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s4WpFjodl2NW876Yw:-K9s4WpFjodl2NW876Yx:-98955813) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L538)):*
bq.eventLog?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[storage/allocator.go, line 125 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s66RvrQuH2qM2P8Is:-K9s66RwSxmO-VDE0_b9:1052869285) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/allocator.go#L125)):*
Error types are named with `Error`, not `Err`

---

*[storage/queue.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nyuL1gaQ20NHx4NkX:-K9s248oQg424qdmSxSN:1107905398) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L162)):*
https://github.com/cockroachdb/cockroach/blob/20e68e7608cd0e227ae797ac7ab5b73d01177a5a/storage/store.go#L281

---

*[storage/queue.go, line 365 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9o--crgomEio7m7nJ3:-K9s3VeR6fF6CjHHSv0M:1688092582) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L365)):*
Yeah, it's been surprisingly complex to get tests to shut down correctly. But if we didn't then it would be harder to detect leaks, and old tests' endlessly-retrying goroutines would make our tests even less deterministic than they are.

 I think it would be a good idea to set this to nil and fix whatever breaks, but don't add it in to this change.

---

*[storage/queue.go, line 429 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s4jMug2VfOsLiGtME:-K9s4jMvzlTuTl2F9Jsw:-658689991) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L429)):*
I think that reusing `CanRetry` for purgatory will lead to problems.  Ranges in purgatory won't be retried until the queue writes to its purgatory channel, which could have the effect of preventing retries if the cause of failure is unrelated to the thing that triggers the purgatory channel. For example, I think `ChangeReplicas` can fail (retryably) if we lose our leadership, in which case the range will stay in purgatory until some (unrelated) gossip update occurs. 

We should use a more precise mechanism for placing ranges in purgatory. In this case, we only want allocatorErrors, so we could make a `purgatoryError` interface that it would implement, or we could remove this decision from the base queue and have the specific queue be the one to call `maybeAddToPurgatory`.

---

*[storage/queue.go, line 440 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s6eJjA2EgLRGQ3iq5:-K9s6eJkfnaTYFZwGamR:2093971472) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L440)):*
The limited size of purgatory is surprising. A store already keeps information about all its ranges in memory at once, so allowing all ranges to be added to purgatory seems reasonable. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[storage/allocator.go, line 128 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s2YDgS1EbRZ_4xFlS:-K9viP6Ueq11ovMojEpV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/allocator.go#L128)):*
Done.

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9vjoxXgVy3gA0Itpq8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
Done.

---

*[storage/queue.go, line 418 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s2QySRRS0RD_oseZr:-K9vkk9pv7KdHnsR5m3W:-397775518) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L418)):*
Error logging is done from the callers. One of the points of this change is to reduce repetitive logging, so we log on the first failure only, not when being retried from purgatory.

---

*[storage/queue.go, line 429 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s4jMug2VfOsLiGtME:-K9vnzvtt6YW-GXLQJBb:1021113265) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L429)):*
Good point. I moved to a `purgatoryError`.

---

*[storage/queue.go, line 440 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s6eJjA2EgLRGQ3iq5:-K9vo2YMTXSqdwik79S0:220607721) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L440)):*
OK, removed the size limitation.b

---

*[storage/queue.go, line 482 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9s4DI9Ocl3no_N0OWb:-K9voDr17XGGWDwcptkR:-646199128) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd558280bfccfe0ae6288ff1e7782d9309b0e104/storage/queue.go#L482)):*
Well whoever added the eventLog change didn't grace the queueLog with any comments, nor did they choose to rewrite these log statements in the first place. There are other pre-existing examples in the file where `log` is used directly, both in what look like traceable code paths and more simple informational logging. It's confusing what should and shouldn't be in the trace log.

I'll move these to use the event log but it's far from clear what the correct course of action is based on existing code and comments.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9vwMy14Tpn9As5vO4D:-76403347) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
still s/to/that/

---

*[storage/queue.go, line 441 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9vxHt4Q62aD5OSi7hg:-K9vxHt4Q62aD5OSi7hh:-1096730560) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1607b07865a904c5ca876d067e428e7f7f2884/storage/queue.go#L441)):*
don't you still want to check `purgatoryError`'s `CanRetry()`?

---

*[storage/queue_test.go, line 476 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9vvoGt5wRhMWilv4rk:-K9vvoGuUPINiXHXdfFF:1337425669) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1607b07865a904c5ca876d067e428e7f7f2884/storage/queue_test.go#L476)):*
isn't this the queue size rather than purgatory size?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9wcX8khecF9ppHUz5M:480488820) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
That doesn't make any sense.

---

*[storage/queue_test.go, line 476 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9vvoGt5wRhMWilv4rk:-K9wcjloX0Jo8y1K2YVA:-1761700750) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1607b07865a904c5ca876d067e428e7f7f2884/storage/queue_test.go#L476)):*
I'm referring to the number of replicas to add to purgatory. It also happens to be the queue size, so there's enough room to rapidly add them to the queue.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9wdCXN5SY5vInP9wtR:598436897) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
The caller of this method does not signal the channel, so i was saying that the comment should read "a channel THAT signals..." though it should probably be the passive "a channel that is signalled"

---

*[storage/queue_test.go, line 476 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9vvoGt5wRhMWilv4rk:-K9wdgS2jsxHA63C28pc:-1820799820) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1607b07865a904c5ca876d067e428e7f7f2884/storage/queue_test.go#L476)):*
OK. Confusing name for a variable used in the maxSize position on the next line.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/queue.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9nzgTcTKp-MrIQOH1V:-K9wiLxAcyqfH-r_S1UH:1283747139) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4f0b7f8a7c840e2d8c7837a69bfd71cd1bd3d8/storage/queue.go#L117)):*
OK. Ambiguous suggestion.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 3 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/allocator.go, line 144 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9wpKh1XTsiSDKUzXqI:-K9wpKh2lfxC51WRnTop:-646997083) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4d6fd0d2e6f5aa7b4fc3c8ba19d76dfdbc232f4/storage/allocator.go#L144)):*
It would be good to have a static `var` assertion that `allocatorError` implements `purgatoryError`.

---

*[storage/queue.go, line 47 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9wouo4Bo9cJpbjFmxz:-K9wouo4Bo9cJpbjFmy-:1741929314) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4d6fd0d2e6f5aa7b4fc3c8ba19d76dfdbc232f4/storage/queue.go#L47)):*
When would a `purgatoryError` not be retryable?  I think this could be a dummy method `purgatoryErrorMarker()` which returns nothing and need not even be called but just serves to implement the interface.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-:-K9wpb40UUfKpsM_lKuC:676834233)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 3 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/queue.go, line 47 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9wouo4Bo9cJpbjFmxz:-K9wwtt5Op1Nus8j2duV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4d6fd0d2e6f5aa7b4fc3c8ba19d76dfdbc232f4/storage/queue.go#L47)):*
Done.

---

*[storage/queue_test.go, line 476 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9vvoGt5wRhMWilv4rk:-K9wx-WtU0Ea4DCBsUut:1614423471) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1607b07865a904c5ca876d067e428e7f7f2884/storage/queue_test.go#L476)):*
Renamed to `replicaCount`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r6.
Review status: 7 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[storage/queue.go, line 193 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9wje2foISmyUATHIM8:-K9wje2g8KHI_HKXMUaH:-688878737) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4d6fd0d2e6f5aa7b4fc3c8ba19d76dfdbc232f4/storage/queue.go#L193)):*
baseQueue.mu was a sync.Locker so that the queue could be returned and embedded in the queueImpl by value. It's probably not worthwhile to keep that, but this method should s/make/new/ to reflect the new return type.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/allocator.go, line 145 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172#-K9x37AI849NmGwFUwsU:-K9x37AI849NmGwFUwsV:2086981513) ([raw file](https://github.com/cockroachdb/cockroach/blob/54850379bce9feced51025a289a13a5b6c0f29c5/storage/allocator.go#L145)):*
this should be a pointer receiver to avoid the case of `allocatorError{}` implementing `purgatoryError` but not `error`. Also, the `purpogatoryError` interface should include `error`, and the comments on this implementation and on the implementation of `Error` above aren't very useful and should be removed, since golint does not require them.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4172)*
<!-- Sent from Reviewable.io -->
